### ISSUE TITLE
SAMPLED_ADDMM

### ISSUE THREAD

**Issue by sunweilunformatech** [tensorflow/tensorflow#56311]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Feature Request

### Source

binary

### Tensorflow Version

2.8

### Custom Code

No

### Python version

3.8

### Current Behaviour?

```shell
Hi TF team,

Is there a plan to add dense-dense matrix multiplication at sparse locations? Pytorch has such support: https://pytorch.org/docs/1.11/generated/torch.sparse.sampled_addmm.html

It is quite useful in ML modules where using sparse connection accelerates the model and helps prevent overfitting. For example, a multi-head attention module can be more efficient if in certain cases we already know that a sparse attention is needed. Currently, the same effect can be achieved by adding the "attention_mask". But there's no performance benefit when the mask is sparse.
```

### Standalone code to reproduce the issue

```shell
layers.MultiHeadAttention(num_heads=num_heads, key_dim=projection_dim, dropout=0.1)(x, x, attention_mask=mask)

Ideally, mask can be both sparse or dense tensor and there's performance improvement when mask's non-zero elements is sparse.
```

### Relevant log output

_No response_</details>

**Comment by sushreebarsa** [comment]
@sunweilunformatech 
Could you elaborate about your feature and please specify the use cases for this feature. Thank you!

**Comment by sunweilunformatech** [comment]
The feature is 2 dense matrix multiplication but only at sparse locations. The result is a sparse matrix. Pytorch already has an API for it [here](https://pytorch.org/docs/1.11/generated/torch.sparse.sampled_addmm.html). The use case is for example keras multi-head attention layer supports a dense attention_mask to mask out some attention connections. This mask should come with some computation speed up in theory as there's no need for the masked element to be computed. Effectively, this requires an API for 2 dense matrix multiplications at sparse locations specified by a sparse matrix.

**Comment by sunweilunformatech** [comment]
One specific example is I wanted to add a multi-head attention layer for vision tasks but only look at a sliding window (like a convolution) for attention. So for a h*w feature map, the full attention's complexity is O(h^2*w^2) which is too expensive for large image size. While in theory, with a k*k sliding window, the complexity can be reduced to O(h*w*k^2).

**Comment by cantonios** [comment]
I don't think we have any current plans for this.  The pytorch version seems to only be available for cuda.  If added to TF, we would want support for cpu/gpu/tpu/xla.  Contributions are welcome in the form of a PR if someone is interested.

**Comment by CohenAriel** [comment]
I'd like to try implementing that.
